lektor/admin/static/js/widgets/ToastEditor.jsx

```diff
@@ -0,0 +1,213 @@
+    let currentEditor = editor.getCurrentModeEditor()
+    let editorHeight
+    try {
+      // markdown
+      let editorEl = currentEditor.editorContainerEl.getElementsByClassName('CodeMirror-sizer')[0]
+      editorHeight = this.outerHeight(editorEl)
+    } catch (e) {
+      // wysiwyg
+
+      // get height of all children in editor
+      let editorEl = currentEditor.$editorContainerEl[0].firstChild
+      let editorChildren = Array.from(editorEl.children)
+      let editorChildrenHeights = editorChildren.map(el => this.outerHeight(el))
+      editorHeight = editorChildrenHeights.reduce((a, b) => a + b)
+
+      // add on padding/border/margin of editor element
+      let marginTop = parseInt(window.getComputedStyle(editorEl).marginTop)
+      let marginBottom = parseInt(window.getComputedStyle(editorEl).marginBottom)
+      let paddingTop = parseInt(window.getComputedStyle(editorEl).paddingTop)
+      let paddingBottom = parseInt(window.getComputedStyle(editorEl).paddingBottom)
+      let borderTop = parseInt(window.getComputedStyle(editorEl).borderTopWidth)
+      let borderBottom = parseInt(window.getComputedStyle(editorEl).borderBottomWidth)
+      editorHeight += (marginTop + marginBottom + paddingTop + paddingBottom + borderTop + borderBottom)
+    }
+
+    let totalHeight = editorHeight + 31 + 31
+    let totalHeightStr = totalHeight + 'px'
+    editor.height(totalHeightStr)
+  }
+
+  onChange () {
+    if (!this.state.editor) {
+      return
+    }
+
+    const markdown = this.state.editor.getMarkdown()
+    if (this.count === 0 && markdown === '') {
+      this.count++
+      return
+    }
+
+    // recalculate height - delay is required to get most up to date height
+    setTimeout(function () {
+      this.recalculateHeight(this.state.editor)
+    }.bind(this), 10)
+
+    // send markdown up
+    let contentChanged = markdown !== this.content
+    if (this.props.onChange && this.content !== null && contentChanged) {
+      this.content = markdown
+      this.props.onChange(markdown)
+    }
+  }
+
+  // focus and blur are used to implement the border-blur found on other widgets
+  focus () {
+    this.setState({
+      style: {
+        borderColor: '#807177',
+        outline: 0,
+        boxShadow: 'inset 0 1px 1px rgba(0,0,0,.075), 0 0 8px rgba(128, 113, 119, 0.6)',
+        transition: 'border-color ease-in-out .15s, box-shadow ease-in-out .15s'
+      }
+    })
+    this.element.firstChild.style.transition = 'border-color ease-in-out .15s, box-shadow ease-in-out .15s'
+    this.element.firstChild.style.borderColor = '#807177'
+  }
+  blur () {
+    this.setState({
+      style: {
+        transition: 'border-color ease-in-out .15s, box-shadow ease-in-out .15s'
+      }
+    })
+    this.element.firstChild.style.borderColor = '#e5e5e5'
+  }
+
+  onRef (element) {
+    if (element === null) {
+      return
+    }
+    this.element = element
+    this.content = this.props.value
+    let editor = new Editor({
```

Comment:
Is there a reason for not using `componentDidMount` and `componentWillUmount` to handle the creation of the editor? If you do, the editor doesn't have to be part of the state.

Reply:
I didn't use `componentDidMount` because I wasn't aware of it (I've learnt React as I've added this widget in).

I've moved it to `componentDidMount` in 6447cbd which makes more sense, but kept it in state, as when the editor is created it needs to be passed to the two widgets, for which a re-render is needed. If there's a better way of doing this then let me know.